src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -4694,14 +4831,58 @@ void StorageReplicatedMergeTree::startupImpl(bool from_attach_thread)
 }
 
 
-void StorageReplicatedMergeTree::partialShutdown()
+void StorageReplicatedMergeTree::flushAndPrepareForShutdown()
+{
+    if (shutdown_prepared_called.exchange(true))
+        return;
+
+    session_expired_callback_handler.reset();
+    stopOutdatedDataPartsLoadingTask();
+
+    /// Cancel fetches, merges and mutations to force the queue_task to finish ASAP.
+    fetcher.blocker.cancelForever();
+    merger_mutator.merges_blocker.cancelForever();
+    parts_mover.moves_blocker.cancelForever();
+    mutations_finalizing_task->deactivate();
+    stopBeingLeader();
+
+    if (attach_thread)
+        attach_thread->shutdown();
```

Comment:
Hm, looks like we only need this (not sure about outdated parts and attach_thread), everything else can be stopped in normal `shutdown`.

However, `mutations_finalizing_task` is suspicious: we deactivate it before stopping restarting_thread, but restarting_thread can restart it: 
https://github.com/ClickHouse/ClickHouse/blob/2924837c7e6030058b87b9a65a06276543430cfd/src/Storages/MergeTree/ReplicatedMergeTreeRestartingThread.cpp#L144

(we already had this issue before)